### PR TITLE
add support for github integrations

### PR DESCRIPTION
**Pull request by macqueen**
The biggest outstanding thing left to do with this is figure out how we're going to migrate existing users. The GitHub apps have Oauth included, but the endpoints that are available are [severely limited](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/identifying-users-for-github-apps/#user-to-server-requests) and don't include all the endpoints we need for our GitHub issues integration or the commit fetching for our current version of releases with GitHub.

So, I think we have to support managing both an Oauth app in GitHub and the new Integrations version at first and then we can add integrations support to the issue tracking plugin and start trying to get customers to move over.

Thoughts?

plugins changes are here for reference: https://github.com/getsentry/sentry-plugins/pull/188

### CODE REVIEW

File: src/sentry/static/sentry/app/views/organizationInstallations.jsx
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/organizationInstallations.jsx]: lgtm

**Comment by dcramer** [comment]
Yeah we'll need both. The integration itself just seems a better way to let us manage webhooks/overall metadata. We'll still need to auth on behalf of users for issue creation.

**Comment by macqueen** [comment]
This is getting to the point where I could use some review.

As is, flow would be:
1. Go to an Installations page for an organization
2. User is prompted to link Github account
3. User is linked to github to install our integration on org of their choice
4. Within Github UI, they choose which repos to include in the integration (they can change this later)
5. They return to installations page (this part of flow needs more work. right now they'd have to go back themselves, but i can add a redirect view)
6. They chose which github org to link to their sentry org, and then we automatically create the repositories that they've given us access to
7. When they create releases, they'd no longer need to have a user token (I just need to fix the validation for this)

There is still a pretty long list of things i need to do:
- [ ] write tests
- [ ] fix validation for repo creation so we check if repo has installation
- [ ] handle moving existing repos to installations
- [ ] decide what to do when repos are removed from installations
- [ ] decide what to do when we're completely uninstalled
- [ ] probably other things

I will definitely write tests and fix the repo validation, but I'd argue for merging this feature gated only to us before handling other things if that's ok with people. I think it'd make it a lot easier to hammer out flow if this were live and something people could try out. This also will just be massive if I wait until everything is perfect and I don't want this to be even worse to review than it already is.

**Comment by macqueen** [comment]
ok @dcramer @mattrobenolt this is ready for review. currently, it's on a page that is not accessible via navigation, so it wouldn't actually be available to users yet. There are still some UI issues, but I'd like to get this in as is so it's easier to review and so I can possibly enlist some other people in helping out with UI stuff.

this one also needs review: https://github.com/getsentry/sentry-plugins/pull/188

I will also need to update a bunch of settings variables/get help from one of you to actually set up our sentry integration in github.

File: src/sentry/api/serializers/models/provider.py
Comment (mattrobenolt) [inline src/sentry/api/serializers/models/provider.py]: You could/should also add `.distinct()` here, which will dedupe on the database side before it's returned.

File: src/sentry/models/installation.py
Comment (mattrobenolt) [inline src/sentry/models/installation.py]: Is it actually possible that these will be NULL? If so, now's the time to adjust since it's a new table.

File: src/sentry/plugins/base/bindings.py
Comment (mattrobenolt) [inline src/sentry/plugins/base/bindings.py]: This can just be:

```python
def all(self):
  return six.iteritems(self._items)
```

You're just repeating what it's already doing.

File: src/sentry/api/endpoints/organization_installations.py
Comment (mattrobenolt) [inline src/sentry/api/endpoints/organization_installations.py]: nit: I'd just add another newline above this so it doesn't run into the `try..except`

File: src/sentry/models/installation.py
Comment (macqueen) [inline src/sentry/models/installation.py]: this was my way of trying to make it flexible for things other than github, but you're right and i should probably just assume these are always going to be necessary

File: src/sentry/plugins/base/bindings.py
Comment (macqueen) [inline src/sentry/plugins/base/bindings.py]: oops

File: src/sentry/api/endpoints/organization_installations.py
Comment (dcramer) [inline src/sentry/api/endpoints/organization_installations.py]: this could be slow if theres multiple providers, but we can probably overlook it for now

File: src/sentry/api/serializers/models/provider.py
Comment (dcramer) [inline src/sentry/api/serializers/models/provider.py]: seems like you'd want this to limit to provider

File: src/sentry/models/installation.py
Comment (dcramer) [inline src/sentry/models/installation.py]: should we add some kind of config-like field to future proof this?

also im not sure this would work with bitbucket

e.g. id like to be able to say "this bitbucket authorization is associated with the org"

More or less, I think this should operate similar to UserSocialAuth/AuthIdentity, but instead of being coupled to a user, its coupled to an org. Additionally when possible it uses an org-wide auth flow (e.g. github integrations), and when not, uses a user-wide flow (e.g. bitbucket?)

File: src/sentry/static/sentry/app/views/organizationInstallations.jsx
Comment (dcramer) [inline src/sentry/static/sentry/app/views/organizationInstallations.jsx]: Maybe "Link Another Account" or Identity ?

File: src/sentry/static/sentry/app/views/organizationInstallations.jsx
Comment (dcramer) [inline src/sentry/static/sentry/app/views/organizationInstallations.jsx]: text needs updated

**Review by dcramer — COMMENTED** [review]
i still have some questions about things, but did an initial pass

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR includes migrations</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/social_auth/backends/github_apps.py


## Migration Checklist

- [ ] new columns need to be nullable (unless table is new)
- [ ] migration with any new index needs to be done concurrently
- [ ] data migrations should not be done inside a transaction
- [ ] before merging, check to make sure there aren't conflicting migration ids


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by macqueen** [comment]
ok @dcramer this is at a point where I could use your review again. Overview:
- supports both old github oauth and new github apps
- allows users to link either a github app auth or a social auth to an organization
- uses either saved auth or the auth of requesting user to fetch commits
- see https://github.com/getsentry/sentry-plugins/pull/188 for how I'm supporting both github apps and oauth

I still need to write a bunch more tests and do some more QA, but would be helpful if you could give it a quick look to make sure it's more or less what you were thinking.

This looks like garbage, but it's the only visual thing I have:
screenshot 2017-07-21 15 18 24

File: src/sentry/api/endpoints/organization_integrations.py
Comment (dcramer) [inline src/sentry/api/endpoints/organization_integrations.py]: maybe just call this 'provider' assuming its a string like 'github'

File: src/sentry/api/endpoints/organization_integrations.py
Comment (dcramer) [inline src/sentry/api/endpoints/organization_integrations.py]: this should be an integer right?

File: src/sentry/api/endpoints/organization_integrations.py
Comment (dcramer) [inline src/sentry/api/endpoints/organization_integrations.py]: this should be an integer?

File: src/sentry/api/serializers/models/provider.py
Comment (dcramer) [inline src/sentry/api/serializers/models/provider.py]: use six.text_type here as javascript fails to handle large integers correctly

File: src/sentry/plugins/providers/base.py
Comment (dcramer) [inline src/sentry/plugins/providers/base.py]: so this always requires a default social auth attached, did we know that its always going to be valid that this constraint is true

e.g. are there any obvious integrations that might not have coupling to a user?

File: src/sentry/plugins/providers/base.py
Comment (dcramer) [inline src/sentry/plugins/providers/base.py]: nbd but maybe this could use a serializer?

File: src/sentry/plugins/providers/base.py
Comment (dcramer) [inline src/sentry/plugins/providers/base.py]: doesnt .first() evaluate immediately?

File: src/sentry/api/endpoints/organization_integrations.py
Comment (dcramer) [inline src/sentry/api/endpoints/organization_integrations.py]: could we put a feature detect test in the post/get endpoints just so we dont have users trying to mess with things before this is in a better state

**Review by dcramer — CHANGES_REQUESTED** [review]

File: src/sentry/plugins/providers/base.py
Comment (macqueen) [inline src/sentry/plugins/providers/base.py]: I figured that not using social auth was probably the exception to the rule, so plugins could override if necessary. this is what i did for the github apps integration: https://github.com/getsentry/sentry-plugins/pull/188/files#diff-038ed6e2271ba416bd35027ef4cf762eR409

Comment (dcramer) [inline src/sentry/plugins/providers/base.py]: alright :+1:

File: src/sentry/api/endpoints/organization_integrations.py
Comment (macqueen) [inline src/sentry/api/endpoints/organization_integrations.py]: done

**Review by dcramer — APPROVED** [review]

File: src/sentry/models/repository.py
Comment (macqueen) [inline src/sentry/models/repository.py]: note to self: create this concurrently before merging

**Comment by macqueen** [comment]
sql for migration:

```sql
   = CREATE TABLE "sentry_organizationintegration" ("id" bigserial NOT NULL PRIMARY KEY, "organization_id" bigint NOT NULL, "integration_id" bigint NOT NULL) []
   = ALTER TABLE "sentry_organizationintegration" ADD CONSTRAINT "sentry_organizationintegr_organization_id_77bd763ea752b4b7_uniq" UNIQUE ("organization_id", "integration_id") []
   = CREATE TABLE "sentry_integration" ("id" bigserial NOT NULL PRIMARY KEY, "provider" varchar(64) NOT NULL, "external_id" varchar(64) NOT NULL, "config" text NOT NULL, "name" varchar(200) NOT NULL, "default_auth_id" integer CHECK ("default_auth_id" >= 0) NULL) []
   = ALTER TABLE "sentry_integration" ADD CONSTRAINT "sentry_integration_provider_e9944c77818d5f5_uniq" UNIQUE ("provider", "external_id") []
   = CREATE TABLE "sentry_projectintegration" ("id" bigserial NOT NULL PRIMARY KEY, "project_id" bigint NOT NULL, "integration_id" bigint NOT NULL, "config" text NOT NULL) []
   = ALTER TABLE "sentry_projectintegration" ADD CONSTRAINT "sentry_projectintegration_project_id_b772982487a62fd_uniq" UNIQUE ("project_id", "integration_id") []
   = ALTER TABLE "sentry_repository" ADD COLUMN "integration_id" integer CHECK ("integration_id" >= 0) NULL; []
   = ALTER TABLE "sentry_repository" ALTER COLUMN "integration_id" TYPE integer, ADD CONSTRAINT CK_integration_id_PSTV_4f3ef70e3e282bab CHECK ("integration_id" >= 0), ALTER COLUMN "integration_id" DROP NOT NULL, ALTER COLUMN "integration_id" DROP DEFAULT; []
   = CREATE INDEX CONCURRENTLY sentry_repository_integration_id ON sentry_repository (integration_id) []
   = ALTER TABLE "sentry_organizationintegration" ADD CONSTRAINT "organization_id_refs_id_af8cad03" FOREIGN KEY ("organization_id") REFERENCES "sentry_organization" ("id") DEFERRABLE INITIALLY DEFERRED; []
   = CREATE INDEX "sentry_organizationintegration_organization_id" ON "sentry_organizationintegration" ("organization_id"); []
   = ALTER TABLE "sentry_organizationintegration" ADD CONSTRAINT "integration_id_refs_id_fdcbef56" FOREIGN KEY ("integration_id") REFERENCES "sentry_integration" ("id") DEFERRABLE INITIALLY DEFERRED; []
   = CREATE INDEX "sentry_organizationintegration_integration_id" ON "sentry_organizationintegration" ("integration_id"); []
   = CREATE INDEX "sentry_integration_default_auth_id" ON "sentry_integration" ("default_auth_id"); []
   = ALTER TABLE "sentry_projectintegration" ADD CONSTRAINT "project_id_refs_id_41efed36" FOREIGN KEY ("project_id") REFERENCES "sentry_project" ("id") DEFERRABLE INITIALLY DEFERRED; []
   = CREATE INDEX "sentry_projectintegration_project_id" ON "sentry_projectintegration" ("project_id"); []
   = ALTER TABLE "sentry_projectintegration" ADD CONSTRAINT "integration_id_refs_id_13d343b7" FOREIGN KEY ("integration_id") REFERENCES "sentry_integration" ("id") DEFERRABLE INITIALLY DEFERRED; []
   = CREATE INDEX "sentry_projectintegration_integration_id" ON "sentry_projectintegration" ("integration_id"); []
````

**Pull request merged**
